File: backend/dto/article.go
// articleのバリデーション/リクエスト関連処理
package dto

import (
	"time"
)

// 定数の定義
const (
	defaultArticleLimit         int = 10
	defaultArticleOffset        int = 0
	DefaultTrendingArticleLimit int = 30
)

// ArticleRequest はリクエストパラメータの構造体
type ArticleRequest struct {
	Category string `json:"category"`
	Tag      string `json:"tag"`
	Limit    int    `json:"limit" validate:"min=1,max=100"`
	Offset   int    `json:"offset" validate:"min=0"`
	Keyword  string `json:"keyword"`
}

func (r *ArticleRequest) SetDefaults() {
	if r.Limit <= 0 {
		r.Limit = defaultArticleLimit
	}
	if r.Offset < 0 {
		r.Offset = defaultArticleOffset
	}
}

// Article represents the structure of an article
type ArticleResponse struct {
	ID          int64     `json:"id"`
	Feed        string    `json:"feed"`
	Category    string    `json:"category"`
	Tags        []string  `json:"tags"`
	Title       string    `json:"title"`
	Link        string    `json:"link"`
	Description *string   `json:"description"`
	PubDate     time.Time `json:"pub_date"`
	ImageURL    string    `json:"image_url"`
	CreatedAt   time.Time `json:"created_at"`
	UpdatedAt   time.Time `json:"updated_at"`
}

// Meta represents the metadata for the response
type Meta struct {
	Total          int    `json:"total"`
	Limit          int    `json:"limit"`
	Offset         int    `json:"offset"`
	RankingEventID string `json:"ranking_event_id"`
}

// Response represents the structure of the API response
type Response struct {
	Data []ArticleResponse `json:"data"`
	Meta Meta              `json:"meta"`
}

File: frontend/src/entities/article.ts
import sanitizeHtml from 'sanitize-html';

export class Article {
    id: number;
    category: string;
    tags: string[];
    title: string;
    link: string;
    description: string;
    feed: string;
    pub_date: Date;
    image_url: string;
    created_at: Date;
    updated_at: Date;

    constructor(data: Partial<Article>) {
        this.id = data.id ?? 0;
        this.category = data.category ?? '';
        this.tags = data.tags ?? [];
        this.title = data.title ?? '';
        this.link = data.link ?? '';
        this.description = data.description ?? '';
        this.feed = data.feed ?? '';
        this.pub_date = data.pub_date ? new Date(data.pub_date) : new Date();
        this.image_url = data.image_url ?? '';
        this.created_at = data.created_at ? new Date(data.created_at) : new Date();
        this.updated_at = data.updated_at ? new Date(data.updated_at) : new Date();
    }

    static fromApiResponse(response: any): Article {
        // descriptionのHTMLタグを除去
        const description = response.description || '';
        const sanitizedDescription = sanitizeHtml(description, {
            allowedTags: [], // すべてのタグを許可しない
        });

        return new Article({
            id: response.id,
            category: response.category,
            tags: response.tags,
            title: response.title,
            link: response.link,
            description: sanitizedDescription,
            feed: response.feed,
            pub_date: response.pub_date,
            image_url: response.image_url,
            created_at: response.created_at,
            updated_at: response.updated_at,
        });
    }
}

